libs/community/langchain_community/chains/pebblo_retrieval/utilities.py

```diff
@@ -334,6 +354,150 @@ async def acheck_prompt_validity(
                 prompt_entities["entityCount"] = pebblo_resp.get("entityCount", 0)
         return is_valid_prompt, prompt_entities
 
+    def is_privileged_user(self, auth_context: Optional[AuthContext]) -> bool:
+        """
+        Check if the user is a privileged user.
+
+        Args:
+            auth_context (Optional[AuthContext]): Authentication context.
+
+        Returns:
+            bool: True if the user is a privileged user, False otherwise.
+        """
+        if not auth_context or not self.policy_cache:
+            return False
+
+        # Get privileged_identities from the policy
+        privileged_identities = self.policy_cache.privileged_identities
+        if not privileged_identities:
+            logger.debug("Privileged identities not found in the policy.")
+            return False
+
+        # Check if user is a privileged user
+        user_auth = auth_context.user_auth if auth_context.user_auth else []
+        is_privileged_user = any(
+            _identity in privileged_identities for _identity in user_auth
+        )
+        if is_privileged_user:
+            logger.debug(f"User {auth_context.user_id} is a privileged user.")
+        else:
+            logger.debug(f"User {auth_context.user_id} is not a privileged user.")
+        return is_privileged_user
+
+    def get_semantic_context(
+        self, auth_context: Optional[AuthContext]
+    ) -> Optional[SemanticContext]:
+        """
+        Generate semantic context based on the given auth context.
+
+        Args:
+            auth_context (Optional[AuthContext]): Authentication context.
+
+        Returns:
+            Optional[SemanticContext]: Semantic context.
+        """
+        semantic_context = None
+        if not auth_context or not self.policy_cache:
+            return semantic_context
+
+        user_semantic_guardrail = self.policy_cache.user_semantic_guardrail
+        if not user_semantic_guardrail:
+            return semantic_context
+
+        _all_guardrails = []
+        user_auth = auth_context.user_auth if auth_context.user_auth else []
+        for identity in user_auth:
+            if guardrail := self.policy_cache.user_semantic_guardrail.get(identity):
+                _all_guardrails.append(guardrail)
+
+        semantic_context = self._combine_all_semantic_filters(_all_guardrails)
+        return semantic_context
+
+    def _start_policy_refresh_thread(self) -> None:
+        """Start a thread to fetch policy from the Pebblo cloud."""
+        logger.info("Starting policy refresh thread.")
+        policy_thread = threading.Thread(target=self._fetch_policy, daemon=True)
```

Comment:
I don't think starting background threads is desired - why would this be necessary

Reply:
We will consider moving away from using threads and instead implement a cache with TTL